lib/src/client.ts

```diff
@@ -27,220 +26,243 @@ export interface HeadersObject {
-        const json = JSON.parse(body);
-        resource = parseJsonADResource(json, resource, store);
-      } catch (e) {
-        throw new AtomicError(
-          `Could not parse JSON from fetching ${subject}. Is it an Atomic Data resource? Error message: ${e.message}`,
-        );
-      }
-    } else if (response.status === 401) {
-      throw new AtomicError(body, ErrorType.Unauthorized);
-    } else if (response.status === 500) {
-      throw new AtomicError(body, ErrorType.Server);
-    } else if (response.status === 404) {
-      throw new AtomicError(body, ErrorType.NotFound);
-    } else {
-      throw new AtomicError(body);
+  /** Throws an error if the URL is not valid */
+  public static tryValidURL(subject: string | undefined): void {
+    try {
+      new URL(subject as string);
+    } catch (e) {
+      throw new Error(`Not a valid URL: ${subject}. ${e}`);
     }
-  } catch (e) {
-    resource.setError(e);
   }
 
-  resource.loading = false;
-  store && store.addResource(resource);
+  /** Throws an error if the URL is not valid */
+  public static isValidURL(subject: string | undefined): boolean {
+    if (typeof subject !== 'string') return false;
 
-  return resource;
-}
+    try {
+      Client.tryValidURL(subject);
 
-/** Posts a Commit to some endpoint. Returns the Commit created by the server. */
-export async function postCommit(
-  commit: Commit,
-  /** URL to post to, e.g. https://atomicdata.dev/commit */
-  endpoint: string,
-): Promise<Commit> {
-  const serialized = serializeDeterministically({ ...commit });
-  const requestHeaders: HeadersInit = new Headers();
-  requestHeaders.set('Content-Type', 'application/ad+json');
-  let response;
-
-  try {
-    response = await fetch(endpoint, {
-      headers: requestHeaders,
-      method: 'POST',
-      body: serialized,
-    });
-  } catch (e) {
-    throw new AtomicError(`Posting Commit to ${endpoint} failed: ${e}`);
+      return true;
+    } catch (e) {
+      return false;
+    }
   }
 
-  const body = await response.text();
+  /**
+   * Removes query params from the URL if it can build a URL. Will return the
+   * normal URL if things go wrong.
+   */
+  // TODO: Not sure about this. Was done because `new Commit()` failed with `unknown-subject`.
+  public static removeQueryParamsFromURL(subject: string): string {
+    return subject?.split('?')[0];
+  }
 
-  if (response.status !== 200) {
-    throw new AtomicError(body, ErrorType.Server);
+  public setFetch(fetchOverride: typeof fetch) {
+    this.__fetchOverride = fetchOverride;
   }
 
-  return parseCommit(body);
-}
+  /**
+   * Fetches and Parses a Resource. Can fetch through another atomic server if you
+   * pass the `from` argument, which should be the baseURL of an Atomic Server.
+   * Returns a tuple of the requested resource and a list of all resources found in the response.
+   */
+  public async fetchResourceHTTP(
+    subject: string,
+    opts: FetchResourceOptions = {},
+  ): Promise<HTTPResult> {
```

Comment:
I feel like returning a nested array like so might be a mistake, and could hurt us later on.

I think returning an object with clearly named parameters is generally better, unless an API is very stable and used all the time (like `useState` in react).

I understand your concern (you often find yourself renaming returned object keys), but if we now want to make changes to `fetchResourceHTTP`'s return value (e.g. change the meaning of the second returned thing) people who use this might not see any errors while updating the library.

What do you think about this @Polleps? Do you think this risk is manageable?